src/creeps/Zombie.cpp

```diff
@@ -57,38 +60,148 @@ bool Zombie::isMoving() const {
+            case ZombieDirection::DIR_R:
+                newDir = (nextDir == ZombieDirection::DIR_RD ? 
+                          ZombieDirection::DIR_D : ZombieDirection::DIR_U);
+                break;
+            case ZombieDirection::DIR_RD:
+                newDir = (nextDir == ZombieDirection::DIR_D ? 
+                          ZombieDirection::DIR_LD : ZombieDirection::DIR_RU);
+                break;
+            case ZombieDirection::DIR_D:
+                newDir = (nextDir == ZombieDirection::DIR_LD ?
+                          ZombieDirection::DIR_L : ZombieDirection::DIR_R);
+                break;
+            case ZombieDirection::DIR_LD:
+                newDir = (nextDir == ZombieDirection::DIR_L ?
+                          ZombieDirection::DIR_LU : ZombieDirection::DIR_RD);
+                break;
+            case ZombieDirection::DIR_L:
+                newDir = (nextDir == ZombieDirection::DIR_LU ? 
+                          ZombieDirection::DIR_U : ZombieDirection::DIR_D);
+                break;
+            case ZombieDirection::DIR_LU:
+                newDir = (nextDir == ZombieDirection::DIR_U ?
+                          ZombieDirection::DIR_RU : ZombieDirection::DIR_LD);
+                break;
+            case ZombieDirection::DIR_U:
+                newDir = (nextDir == ZombieDirection::DIR_RU ?
+                          ZombieDirection::DIR_R : ZombieDirection::DIR_L);
+                break;
+            case ZombieDirection::DIR_RU:
+                newDir = (nextDir == ZombieDirection::DIR_R ?
+                          ZombieDirection::DIR_RD : ZombieDirection::DIR_LU);
+                break;
+            case ZombieDirection::DIR_INVALID:
+                break;
+        }
+
+        // set current direction
+        setCurDir(newDir);
+    }
 }
 
 /**
  * Get the direction of the zombie and take a step in the appropriate direction
- * Rob, Fred
- * March 13
+ * Robert Arendac, Fred Yang
+ * March 28
 */
 void Zombie::generateMove() {
-    const ZombieDirection direction = getMoveDir();   //Direction zombie is moving
-    //cout << "move dir: " << d << " state: " << state << " Frame: " << frame << endl;
-
-    // Path is empty, shouldn't move
-    if (direction == ZombieDirection::DIR_INVALID || checkTarget()) {
+     // Direction zombie is moving
+    const ZombieDirection direction = getMoveDir();
+
+    // detect surroundings
+    const int collisionObjId = detectObj();
+    
+    // Path is empty or preferrable objects appear in vicinity, prepared to 
+    // switch state.
+    if (direction == ZombieDirection::DIR_INVALID || collisionObjId > 0) {
         if (frame > 0) {
             --frame;
         }
 
-        // Changed to attack state once attack code is ready
-        setState(ZombieState::ZOMBIE_IDLE);
-
+        if (collisionObjId > 5) { // base
+            setState(ZombieState::ZOMBIE_IDLE);
+        } else if (collisionObjId > 2) { // marine, turret, or barricade
```

Comment:
Base is something the zombie is going to attack as well, you can just change this thing to:
```
if (collisionObjId > 2)
    setState(ZombieState::ZOMBIE_ATTACK);